float64 ddx
float64 ddy